pyomo/contrib/solver/solvers/gurobi/gurobi_direct_base.py

```diff
@@ -0,0 +1,529 @@
+            with capture_output(TeeStream(*ostreams), capture_fd=False):
+                gurobi_model, solution_loader, has_obj = self._create_solver_model(
+                    model
+                )
+                options = config.solver_options
+
+                gurobi_model.setParam('LogToConsole', 1)
+
+                if config.threads is not None:
+                    gurobi_model.setParam('Threads', config.threads)
+                if config.time_limit is not None:
+                    gurobi_model.setParam('TimeLimit', config.time_limit)
+                if config.rel_gap is not None:
+                    gurobi_model.setParam('MIPGap', config.rel_gap)
+                if config.abs_gap is not None:
+                    gurobi_model.setParam('MIPGapAbs', config.abs_gap)
+
+                if config.warmstart_discrete_vars:
+                    self._mipstart()
+
+                for key, option in options.items():
+                    gurobi_model.setParam(key, option)
+
+                timer.start('optimize')
+                gurobi_model.optimize(self._callback)
+                timer.stop('optimize')
+
+            res = self._postsolve(
+                grb_model=gurobi_model, solution_loader=solution_loader, has_obj=has_obj
+            )
+        finally:
+            os.chdir(orig_cwd)
+
+            # hack to work around legacy solver wrapper __setattr__
+            # otherwise, this would just be self.config = orig_config
+            object.__setattr__(self, 'config', orig_config)
+            self.config = orig_config
+
+        res.solver_log = ostreams[0].getvalue()
+        end_timestamp = datetime.datetime.now(datetime.timezone.utc)
+        res.timing_info.start_timestamp = start_timestamp
+        res.timing_info.wall_time = (end_timestamp - start_timestamp).total_seconds()
+        res.timing_info.timer = timer
+        return res
+
+    def _get_tc_map(self):
+        if GurobiDirectBase._tc_map is None:
+            grb = gurobipy.GRB
+            tc = TerminationCondition
+            GurobiDirectBase._tc_map = {
+                grb.LOADED: tc.unknown,  # problem is loaded, but no solution
+                grb.OPTIMAL: tc.convergenceCriteriaSatisfied,
+                grb.INFEASIBLE: tc.provenInfeasible,
+                grb.INF_OR_UNBD: tc.infeasibleOrUnbounded,
+                grb.UNBOUNDED: tc.unbounded,
+                grb.CUTOFF: tc.objectiveLimit,
+                grb.ITERATION_LIMIT: tc.iterationLimit,
+                grb.NODE_LIMIT: tc.iterationLimit,
+                grb.TIME_LIMIT: tc.maxTimeLimit,
+                grb.SOLUTION_LIMIT: tc.unknown,
+                grb.INTERRUPTED: tc.interrupted,
+                grb.NUMERIC: tc.unknown,
+                grb.SUBOPTIMAL: tc.unknown,
+                grb.USER_OBJ_LIMIT: tc.objectiveLimit,
+            }
+        return GurobiDirectBase._tc_map
+
+    def _postsolve(self, grb_model, solution_loader, has_obj):
```

Comment:
I would suggest renaming this something more descriptive like `_populate_results`